tuf/ngclient/_internal/trusted_metadata_set.py

```diff
@@ -340,7 +340,15 @@ def update_snapshot(
         logger.debug("Updated snapshot v%d", new_snapshot.signed.version)
 
         # snapshot is loaded, but we raise if it's not valid _final_ snapshot
-        self._check_final_snapshot()
+        if not offline:
+            self._check_final_snapshot()
+        else:
+            snapshot_meta = self.timestamp.signed.snapshot_meta
+            if self.snapshot.signed.version != snapshot_meta.version:
+                raise exceptions.BadVersionNumberError(
+                    f"Expected snapshot version {snapshot_meta.version}, "
+                    f"got {self.snapshot.signed.version}"
+                )
```

Comment:
I usually don't mind duplicating a bit of code but this has a bit of a smell... I think we could avoid this if the whole TrustedMetadataSet instance was "offline": then the lower level code (in this instance _check_final_snapshot()) could have the differing code paths and this code here would not need changes

Reply:
Yea we were going back and forth for a little while on how we should deal with this. We'll modify _check_final_snapshot() to have differing codepaths

Reply:
Thanks for all the feedback! We've made changes to TrustedMetadataSet so that it allows the offline flag to be set within this object and all methods now take advantage of this.